src/main/java/parser/Parser.java

```diff
@@ -0,0 +1,167 @@
+        if (Pattern.matches("^(find\\s+.*)|(.*\\s+find\\s+.*)|(.*\\s+find$)", input)) {
+            if (!Pattern.matches("^find\\s+.*", input)) {
+                throw new DukeException("Action should be at the front");
+            }
+            if (!Pattern.matches("^find\\s+.*", input)) {
+                throw new DukeException("A task number must be provided");
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param input raw input
+     * @return String the lower cased task type is returned (todo/deadline/event)
+     * @throws DukeException when type is not recognized
+     */
+    public static String getType(String input) throws DukeException {
+        String lowerCaseInput = input.toLowerCase();
+        String acceptedTypes = String.format("(%s)", String.join("|", Parser.taskTypes));
+        if (Pattern.matches(
+                String.format(
+                        "^%s\\s+.*|.*\\s+%s$|.*\\s+%s\\s+.*",
+                        acceptedTypes, acceptedTypes, acceptedTypes),
+                lowerCaseInput)) {
+            if (Pattern.matches(String.format("^%s\\s+.*", acceptedTypes), lowerCaseInput)) {
+                return input.split(" ")[0].toLowerCase();
+            } else {
+                throw new DukeException("Please start with event type");
+            }
+        }
+        throw new DukeException("No accepted types present");
+    }
+
+    /**
+     * @param description user input without the type
+     * @return String returns the content
+     * @throws DukeException when content is empty, raise exception
+     */
+    public static String getContent(String description) throws DukeException {
+        Matcher matcher = Pattern.compile("(/by|/at)").matcher(description);
+        int index = matcher.find() ? matcher.start() : -1;
+        if (index == -1 && description.trim().length() > 0) {
+            return description.trim();
+        }
+        String content = description.substring(0, index).trim();
+        if (content.length() > 0) return content;
+        throw new DukeException("Content cannot be empty!");
+    }
+
+    /**
+     * @param description user input without type
+     * @param regex varies on whether it is event or deadline (/at or /by)
+     * @return String raw datetime string
+     * @throws DukeException if no datetime string is provided, raise exception
+     */
+    public static String getDateTime(String description, String regex) throws DukeException {
+        Matcher matcher = Pattern.compile(regex).matcher(description);
+        int index = matcher.find() ? matcher.start() : -1;
+        if (index == -1) {
+            throw new DukeException(String.format("Please provide %s for this event type", regex));
+        }
+        String dateTime = description.substring(index + regex.length()).trim();
+
+        if (dateTime.length() == 0) throw new DukeException("Please provide a time");
```

Comment:
Take note that according to the textbook Coding Standard, throw new DukeException("Please provide a time"); should still be indented within braces even though there is only one statement.